mailpoet/assets/js/src/segments/dynamic/dynamic-segments-filters/fields/woocommerce/purchased-with-attribute.tsx

```diff
@@ -0,0 +1,137 @@
+import { useDispatch, useSelect } from '@wordpress/data';
+import { __ } from '@wordpress/i18n';
+import { useEffect, useRef } from 'react';
+import { Select } from 'common/form/select/select';
+import { ReactSelect } from 'common/form/react-select/react-select';
+import { filter } from 'lodash/fp';
+import { storeName } from '../../../store';
+import {
+  AnyValueTypes,
+  FilterProps,
+  SelectOption,
+  WindowProductAttributes,
+  WooCommerceFormItem,
+} from '../../../types';
+
+export function validatePurchasedWithAttribute(
+  formItems: WooCommerceFormItem,
+): boolean {
+  const purchasedProductWithAttributeIsInvalid =
+    !formItems.operator ||
+    formItems.attribute_id === undefined ||
+    formItems.attribute_term_id === undefined;
+
+  return !purchasedProductWithAttributeIsInvalid;
+}
+
+export function PurchasedWithAttributeFields({
+  filterIndex,
+}: FilterProps): JSX.Element {
+  const segment: WooCommerceFormItem = useSelect(
+    (select) => select(storeName).getSegmentFilter(filterIndex),
+    [filterIndex],
+  );
+  const { updateSegmentFilter } = useDispatch(storeName);
+
+  const productAttributes: WindowProductAttributes = useSelect(
+    (select) => select(storeName).getProductAttributes(),
+    [],
+  );
+
+  const productAttributesOptions = Object.values(productAttributes).map(
+    (attribute) => ({
+      value: attribute.id,
+      label: attribute.label,
+    }),
+  );
+
+  const productAttributeTermsOptionsRef = useRef(null);
+
+  useEffect(() => {
+    if (segment.attribute_id === undefined) {
+      productAttributeTermsOptionsRef.current = null;
+      return;
+    }
+
+    productAttributeTermsOptionsRef.current = productAttributes[
+      segment.attribute_id
+    ].terms.map((term) => ({
+      value: term.term_id,
+      label: term.name,
+    }));
+  }, [segment.attribute_id, productAttributes]);
```

Comment:
Maybe I'm missing something, but to me, it seems that here we want to "cache" the mapped value of product attributes between renders. Am I correct?
I would say that more natural and probably also more efficient way of doing this could using `useMemo` (https://react.dev/reference/react/useMemo)

Could look like this (not tested):
```js

const productAttributeTermsOptions = useMemo(() => {
  if (segment.attribute_id === undefined) {
     return null;
  }
  return productAttributes[
    segment.attribute_id
  ].terms.map((term) => ({
    value: term.term_id,
    label: term.name,
  })
}, [segment.attribute_id, productAttributes]);

Reply:
Thanks for this comment. I wasn't sure what the intention was with the original `useRef` myself. I've refactored this to use `useMemo`.